### DISCUSSION TITLE
[Version 2] Station Logbooks Development

### DISCUSSION THREAD

**Discussion by magicbug** [magicbug/Cloudlog#1131] [General]
Station Logbooks has been one of the most asked for features and something I've wanted but moving QTH has been a kick to get this actually implemented, in a nutshell, "Station Logbooks" will allow you to link Station Locations together, this will be then used for the overall Awards, Logbook, Analytics. I'm designing it so that a logbook can have many Location Profiles and vs versa so (Many-to-Many) this means that with some further code it will be possible to have shared club-style logbooks, although this will need more code changes.

There's still tons to do, I've got the interface in place for creating a station logbook item but now need to add the code for selecting Station Locations and assigning them to the logbook then some code that returns a list of locations that can be used for the main areas in Cloudlog but one step at a time.

![Screenshot from 2021-08-12 21-10-51](https://user-images.githubusercontent.com/84308/129264130-1706bbbd-58c4-4d4f-be7e-6b801af70bbb.png)

Hopefully the above gives you an idea of what's happened so far, I'm working in the station_logbooks branch this will then be merged into the master branch which might mean that master becomes a bit quiet. while I get things figured out and ask @AndreasK79 for help with the awards bits!

Hopefully the above makes some sense and is not too rambly!

**Comment by AndreasK79** [comment]
Great news. Let me know when you need me to pitch in!

**Comment by magicbug** [comment]
Right at a place now where things can be done within the "Station Logbooks" branch of the codebase you'll find there's now Station Logbooks and Station Locations in the dropdown, You'll need to claim Locations to your profile by clicking the button in the locations box, you can then group by creating a logbook then editing and selecting the associated Locations.

There's a model called **logbooks_model** this contains everything to do with Station Logbooks there's a function list_logbook_relationships that returns an array of station locations associated with a logbook it needs a logbook id passed to it I use the one in the session data.

I've been using

```
$CI =& get_instance();
$CI->load->model('logbooks_model');
$logbooks_locations_array = $CI->logbooks_model->list_logbook_relationships($this->session->userdata('active_station_logbook'));
```

Then adding a  `$this->db->where_in('station_id', $logbooks_locations_array);` within the database call instead of just passing one of them.. basically what needs changed is everything in analytics and awards. I'm trying to put a TODO together to keep a track of whats been done at https://github.com/magicbug/Cloudlog/projects/9

@AndreasK79 if you fancy helping a bit with this then always appreciated especially awards stuff so I don't mess it up!

**Reply by AndreasK79** [reply]
Good work! It seems quite straightforward. I did the CQ Award, and it seems it's mostly changing to "station_id in" most places.
Quick Lookup also needs to be fixed. I'll do more awards when I find some time.

**Reply by magicbug** [reply]
Tried to keep it simple, yep will do Quick Lookup tomorrow during the day, might see if I can speed it up a bit too.

I'll try get new setups a bit more straightforward, pleased you don't hate the implementation :)

**Reply by AndreasK79** [reply]
Speed up is always great. See what you can do.

As for the implementation, it was easy enough to implement. When it comes to the GUI part, it needs to be documented good, or else this will raise many questions. I didn't find it too intuitive to use. 

I think this can be solved better. Wouldn't it be better to have logbooks and locations some how merged? 
And more visually be able to see which location belongs to which logbook?
You also need to be able to change the ownership of a location.

**Reply by AndreasK79** [reply]
@magicbug I see Distances Worked needs a rewrite. Right now, it uses the gridsquare for the active station location. That will not be correct when QSOs are fetched across multiple locations. Accumulated Statistics and Days with QSOs are probably just as easy as the others.

**Comment by Werzi2001** [comment]
Hello @magicbug @AndreasK79 ,

just want to tune into this discussion. As I'm mostly doing sota acivations bascially every activation is a different station in Cloudlog. This causes the Analytics and Awards to be not as useful for me as they only refer to the active station (which is just a single activation) and not all of them. I thought about adding a setting that would disable the station filter for the analytics and awards but if I understand it correctly this change will enable that as well (analytics/awards for multiple stations, basically multiple active stations). Right?

Best regards
Werzi2001

**Reply by AndreasK79** [reply]
Hi @Werzi2001.

Yes, you understood this correct. I've fixed the awards code, and several of the analytics code regarding this. 
This has been a highly requested feature, so I think many will be pleased when it's done.

I've done what I can do regarding this, unless Peter has something else in mind that I can do.

Now we just need to wait for Peter to finish up things.

Andreas.

**Reply by magicbug** [reply]
@Werzi2001 Like @AndreasK79 says this will solve it for you, you'll be able to group locations together into "logbooks" and see stats over all of them.. If you want to start a separate thread on SOTA and how we can improve support for it please do.

**Reply by Werzi2001** [reply]
@magicbug Thanks for your reply. I'll test the station logbook branch a bit locally but it looks good at first glance. Just for understanding: The active station is only relevant for creation of new QSOs right? All awards, analytics etc. should use active logbook? About the SOTA stuff I'll do that as soon as I can really say something about. I literally just got my license a month ago and am only just starting to "see what might be missing".

**Reply by magicbug** [reply]
Correct, Active Station purely for logging purposes.

**Comment by magicbug** [comment]
Thanks @AndreasK79's I'm a bit slow last week pains flared up to uncomfortable levels but pressing on slowly.

**Reply by Werzi2001** [reply]
@magicbug Sad to hear that. Wish you all the best.

**Reply by magicbug** [reply]
Long term just hasn't been great recently! quiet this week as I've been roving gridsquares on satellite passes :)

**Comment by Cadair** [comment]
I would love to test this out and give some feedback.

Is this in a place where I can test it and revert back to the master branch if things go sideways? Is there a clean path from many station locations to one "station location"?

**Reply by Cadair** [reply]
@magicbug I swapped to the master branch to tests the fixes for making a new contest and then swapped back and I get some database errors. Does the dev branch just need an update?

**Reply by AndreasK79** [reply]
@Cadair you need to update migrations-table. 

> update migrations set version = 79;

The scripts are trying to run again, since master is on version 71,

**Reply by Cadair** [reply]
Right, I think. So I am clear:

* I start on master, I have db version 71 in the table.
* I swap to `station_locations` it updates the schema to 79 and does the migrations to the db.
* I swap back to master, it *updates the migrations table to 71*?
* I swap back again to `station_locations` and it tries to reapply the migrations and fails?

**Reply by AndreasK79** [reply]
Yes. The scripts aren't written to check if everything exists. This is only a problem if someone changes branches. The number of people doing that, is very low.

**Reply by Cadair** [reply]
Why would swapping back to master decrement the migration number? Surely migrations can only ever go up?

**Reply by AndreasK79** [reply]
I have no idea, but I'm guessing that is how the migration system in Codeigniter is built.

**Reply by Cadair** [reply]
Fair enough, I don't know much about db migrations but getting yourself in a situation where it reapplys them is annoying :grinning:

**Reply by AndreasK79** [reply]
@Cadair It's your lucky day. I've changed the scripts. You shouldn't get any errors when swapping branches now :)

**Reply by Werzi2001** [reply]
@Cadair @AndreasK79 
As far as I can see migrations should be able to go down (not just up). Each migration (see in `application/migrations`) has an `up` and a `down` method. Maybe there is missing something in the `down` part of a migration that causes problems?
Anyhow as the "problem" is fixed now, it might not be worth investigating.

**Reply by AndreasK79** [reply]
@Werzi2001 you are right. Some of the scripts doesn't have anything in down. However, I just swapped branches now. Down is not executed  for those that have anything in down. So it seems the problem is more complicated than that.

But right now, it is solved by checking if things exists. If they do, we skip.

**Reply by Cadair** [reply]
I just swapped from master back to `station_logbooks` and I get the following database error again:

```

A Database Error Occurred

Error Number: 1146

Table 'cloudlog.queries' doesn't exist

ALTER TABLE queries CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci

Filename: migrations/084_fix_collate.php

Line Number: 33
```

What's the safest fix?

**Reply by Cadair** [reply]
I am now very confused. I have migrations version 73 on master. I don't know why this would error out applying 84?

**Reply by AndreasK79** [reply]
Queries is created in migration 71. It is very strange that you get this, but if you update migrations like this:
`update migrations set version = 70;`
All the migration scripts will fire again, and hopefully fix your problem. I just tried this on my local copy on station_logbooks.

**Reply by Cadair** [reply]
That seems to have done the trick. Given the amount of features and development going on in that branch now, what's the plan with regards to merging it to master?

**Reply by AndreasK79** [reply]
That's great to hear. This branch will never be merged into master (I screwed up some commits a while ago, so it can't be merged into master). It will replace master when it's due for release. I don't know when that will be. Only @magicbug can answer that.

It seems to be stable, but I think Peter still wants a few things to be completed before release.

**Reply by Cadair** [reply]
Is master considered "release"?

**Comment by magicbug** [comment]
You should be able to run the station_logbooks branch without any problem, currently running it on my "live" copy todo beta testing its got some quirks but you wont lose any QSO info.

**Reply by Cadair** [reply]
I have checked out the branch and it seems to be working well. I love having all my stats together :heart:.

The only thing I noticed that I didn't understand was the "claim this location" button on the station locations page. I clicked them all just because but it didn't seem to change anything?

**Reply by magicbug** [reply]
When clicking "Claim" it should have just assigned it to your user ID then hid the button.. hopefully that was the case

**Reply by Cadair** [reply]
It did hide the button. What does assigning it to my user id achieve?

**Reply by magicbug** [reply]
Station Logbooks only shows Station Locations assigned to your user_id for some filtering.. aka slowly improving bits to allow multi user todo more things :)

**Reply by Cadair** [reply]
Ah I made a station logbook with all the locations in before I claimed the locations. Was that supposed to work?

**Reply by magicbug** [reply]
Nope! in that case I may have forgotten something :p

**Reply by Nutster2454** [reply]
Hi 2M0OCC,
Just be reading this tread and would like to understand how you are logging your sota's as I have 3 different callsign plus over 40 different sota hills. How is best to go about this with out make a station logbook for every sota summit and for the three different calls???

**Reply by Werzi2001** [reply]
@Nutster2454 
Well I'm not 2M0OCC but am doing quite a lot SOTA as well. So I'll just tell you my view on this :-)

> How is best to go about this with out make a station logbook for every sota summit and for the three different calls???

Sorry to say that but I guess that's exactly as it's usually done (as how I'm doing it). It's also the same on other platforms (e.g. eQSL.cc). I guess the main reason is because every sota activation has a different location and therefore a different grid locator etc.

If you don't want to do that I guess you could just use one station for all our sota acitivations (but still different ones for the callsignes) but you'll "loose" features like distance calculation as these are based on the locator.

**Reply by magicbug** [reply]
Bit off topic but yes theres really no way around multiple Station Locations, however least now they can be grouped, its a bit beyond the scope of this thread but a quick create of a location can be added, just need to get everything else done first..

**Comment by Nutster2454** [comment]
Ok thank you let me know when things are updated to best reconfig my logs.

**Comment by magicbug** [comment]
Got annoyed with the Edit interface on Station Logbooks so designed it 

![Screenshot from 2021-10-12 15-16-51](https://user-images.githubusercontent.com/84308/136974246-e3520a7c-4fad-45b2-98a1-0986d9c929ff.png)

**Reply by AndreasK79** [reply]
Now this is more like what I imagined how it should look like.

**Reply by Cadair** [reply]
Looks awesome!

Also looks like a quick add location button would fit nicely in there sometime in the future ;)

**Reply by magicbug** [reply]
Yep, either a clone button or a model popup box allowing user to set a few details

**Comment by magicbug** [comment]
Outstanding tasks

- [x] Make sure the user only sees there Station Locations in Station Logbook Edit
- [ ] Make sure user only sees there Locations in "Station Locations"
- [ ] Make sure user only sees their Locations for the active logbook when logging a qso in the location drop down

**Reply by Werzi2001** [reply]
@magicbug 
Would love to see this finished. I created a pull request where I worked on some of your points. Hope that helps at least a bit: #1256

**Reply by magicbug** [reply]
It really does help! dev slowed again due to contest season but back on it again :-)

**Comment by AndreasK79** [comment]
@Werzi2001 great work.

I would also like to see this finished, so that we can merge the two branches.

**Comment by magicbug** [comment]
More bits coming along, starting to implement login changes, need to build the public "logbook view", all this needs done to get ready to make things live in the master branch and tagged v2 but it will get there, this branch is however "stable" its my daily logging system codebase at the moment!

**Comment by Cadair** [comment]
I seem to be encountering some weirdness I am struggling to figure out on the station_logbooks branch while pushing ADIF QSOs over the API. The API call is returning success but the entry isn't actually appearing in the logbook anywhere.

**Reply by AndreasK79** [reply]
@Cadair yes, I found the problem. Since the API doesn't have a station_id, the QSO is never saved.

The problem lies in Logbook_model.php: line 1777 to 1779. Since station_id 0 doesn't belong to a user, the function returns.

@magicbug should we remove this? This breaks API-logging.

**Reply by magicbug** [reply]
Sorry been quite, its been major contest season so barely been home! 

If I remember right the API for accepting a QSO also allowed passing a station_id to assign it.. 

My thinking was really

- API POST
- Accepts a station ID or some other ID to say what it is.
- if ID isn't provided API key just checks what the users default one is
- saves QSO.

This will need implementing anyway due to the changes.

**Reply by Cadair** [reply]
As you say the API currently needs a `"station_profile_id"` field.

Having a default would also be a nice change.

**Comment by Cadair** [comment]
Another thing I noticed is that I imported an ADIF file twice, once with one location and once with another and it duplicated it into both locations, which wasn't what I wanted.

I don't know if this is just a case of "feature not bug" or what, but I now have to figure out how to remove a couple of hundred QSOs from my log.

**Reply by AndreasK79** [reply]
I have no idea how that is possible. I tried to reproduce here, but no such luck.

If you are able, and comfortable in running sql in the database, I can give you the necessary queries to fix this.

**Reply by magicbug** [reply]
I cant reproduce that either so somethings very weird going on.

**Reply by Cadair** [reply]
Hmmm, I don't exactly know when this happened, I have been re-importing the WSJT adif. I might have done the import on an older version of the branch :shrug:.

If you could give me some hints on the SQL needed to purge them that would be great.

**Reply by AndreasK79** [reply]
@Cadair here it goes:

1. You need to find the station_id to the two different locations. You can find them under station locations.
2.  If you have another name on the table other than the default, you need to change that in the query.

Query 1. This is so you can inspect the QSOs, and see that you find the correct QSOs.

> select COL_TIME_ON, COL_CALL, COL_BAND, COL_MODE, COL_SUBMODE, COL_PROP_MODE, station_id 
from table_hrd_contacts_v01  
where station_id = :stationidtodeletefrom
and exists (
	select 1 
	from table_hrd_contacts_v01 thcv
	where station_id = :stationidtokeep 
	and COL_CALL = table_hrd_contacts_v01.COL_CALL 
	and COL_BAND = table_hrd_contacts_v01.COL_BAND 
	and COL_TIME_ON = table_hrd_contacts_v01.COL_TIME_ON 
	and COL_MODE = table_hrd_contacts_v01.COL_MODE 
	and COL_SUBMODE = table_hrd_contacts_v01.COL_SUBMODE
	and COL_PROP_MODE = table_hrd_contacts_v01.COL_PROP_MODE);

Query 2. This deletes the duplicate QSOs the query finds.

> delete from table_hrd_contacts_v01 
where station_id = :stationidtodeletefrom
and exists (
	select 1 
	from table_hrd_contacts_v01 thcv
	where station_id = :stationidtokeep
	and COL_CALL = table_hrd_contacts_v01.COL_CALL 
	and COL_BAND = table_hrd_contacts_v01.COL_BAND 
	and COL_TIME_ON = table_hrd_contacts_v01.COL_TIME_ON 
	and COL_MODE = table_hrd_contacts_v01.COL_MODE 
	and COL_SUBMODE = table_hrd_contacts_v01.COL_SUBMODE
	and COL_PROP_MODE = table_hrd_contacts_v01.COL_PROP_MODE);

Make sure you have a backup just in case.

**Comment by magicbug** [comment]
Apologies got being quiet, contest seasons easing off again so have more "free" time to get things back on track

**Comment by Werzi2001** [comment]
@magicbug @AndreasK79 Just wondering what is still missing for V2. I know there is a checklist somewhere but I can't find it anymore. Would love to see this finished and have an official V2.

**Reply by AndreasK79** [reply]
I'm not really sure. I can't find the topic with the list anymore myself, so Pete needs to answer this one.

These are the ones I remember.

- Club Log cron job to upload for all users, without providing username and password.
- eQSL - think it was a cron job to upload here as well
- Replace Google Charts on Analytics and Statistics with other charts (chart.js or highcharts)

**Reply by m0urs** [reply]
If these are the only missing things: Would it be safe to use V2 already in the current state? And if yes, do I need to do anything besides just pulling the "station_logbooks" branch? (I currently have used "Station Locations" already with Version 1)

**Reply by AndreasK79** [reply]
Those were the ones I remembered. A thread with info existed, but it seems it's gone.

It is stable enough to run as a daily. I know @magicbug runs it himself. You need to pull the station_logbook branch, then configure a station logbook and add station locations to that. Other than that, it should work ok.

**Reply by m0urs** [reply]
> Club Log cron job to upload for all users, without providing username and password.

Is Clublog upload in general not available yet in version 2? Or do I need to modify my existing cronjob I was using with the old version
`curl --silent https://cloudlog.urspringer.de/index.php/clublog/upload/dg3nab`
Because this no longer seems to work after switching to version 2

I also would say, that updating the clublog SCP database is no longer working. At least the last update date was the one before I switched to version 2. I am using this cronjob:
`curl --silent https://cloudlog.urspringer.de/index.php/update/update_clublog_scp `

**Comment by m0urs** [comment]
Thank you. Do you know if I would be able to switch back in case something
important is not working as expected or is something altered in the
database structure after switching to the new version?

Andreas Kristiansen ***@***.***> schrieb am So. 26. Juni 2022
um 19:14:

> Those were the ones I remembered. A thread with info existed, but it seems
> it's gone.
>
> It is stable enough to run as a daily. I know @magicbug
> <https://github.com/magicbug> runs it himself. You need to pull the
> station_logbook branch, then configure a station logbook and add station
> locations to that. Other than that, it should work ok.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1131#discussioncomment-3026895>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATX5SVQTVH7ZGDFHPQC65DVRCFVRANCNFSM5CCCR36Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
https://urspringer.de/info
Signal: <https://signal.org> +4915143106798
PGP-Verschlüsselung <https://www.gpg4win.de/>: https://urspringer.de/pgp

**Reply by AndreasK79** [reply]
A lot has been changed in the database, but switching should not be a problem. I tried on my local machine to switch back to master, and it worked here.

As always, make a backup, just in case. Better safe than sorry,

**Reply by m0urs** [reply]
I was not really successful. After switching to the station_logbook branch I did get some database related PHP errors. So I guess that I would have needed to adapt the db structure to the new version. I then decided to start with a new database and imported all my QSOs. This in general worked but I lost my QSL card images as the QSO numbers have changed (I guess) during the ADIF import. 

I liked what I saw with the new version, but then I decided to switch back and wait until the official release of version 2 :-) (and I hope that then an official migration guide is provided :-))

**Reply by AndreasK79** [reply]
Sorry to hear that. Those errors would have been nice to know about, as things should work when switching over. You don't happen to have them in a log somewhere?

**Reply by m0urs** [reply]
I tried it again to reproduce the error. This is what I did:

`git checkout station_logbooks`

Now I started Cloudlog. It did not show any QSOs on the start page.

Now I clicked on "Logbook".

Then the following error occured:

```
An uncaught Exception was encountered
Type: TypeError

Message: Argument 2 passed to CI_DB_query_builder::where_in() must be of the type array, bool given, called in /var/www/cloudlog.urspringer.de/html/application/models/Logbook_model.php on line 1025

Filename: /var/www/cloudlog.urspringer.de/html/system/database/DB_query_builder.php

Line Number: 720

```
Then I switched back with 

`git checkout master`

Everything was fine again.

**Reply by phl0** [reply]
> Then the following error occured:
> 
> ```
> An uncaught Exception was encountered
> Type: TypeError
> 
> Message: Argument 2 passed to CI_DB_query_builder::where_in() must be of the type array, bool given, called in /var/www/cloudlog.urspringer.de/html/application/models/Logbook_model.php on line 1025
> 
> Filename: /var/www/cloudlog.urspringer.de/html/system/database/DB_query_builder.php
> 
> Line Number: 720
> ```

The root cause is this:  https://github.com/magicbug/Cloudlog/issues/1481

**Reply by m0urs** [reply]
Thank you. You are right, after I defined a station logbook and added my locations to it, the error went away and my QSOs are displayed now.

I then had the issue that none of my Notes appeared. The reason for that is that the field "user_id" in the table "notes" is "NULL" for all entries. After manually changing it to the user ID of my user (in my case "4"), all Notes will now be displayed.

The same issue was with the tables "api", "cat". I changed the "user_id" here as well for all entries.

Now, at the first sight, everything works again.

I am still not sure if I should stay with it for production use ;-)

**Reply by AndreasK79** [reply]
Thanks @m0urs and @phl0. I'll look into that error later. Regarding userid, nothing is set in the migration scripts. If only one user is registered, it is easy, but if several exists in the database, how do we know which one to pick and set? 

@m0urs I say stay. Since you got it working, no reason to go back.

**Reply by m0urs** [reply]
Ok, I will give it a try.

Can the debugging information in the QSO list somehow be deactivated? It is not really an issue but if there is an easy way to do it .. ?

Regarding the user_id: We did have more than one user already in the old version, am I right? So how was it done there? Or was this data valid for all users before and is now on a per user base?

**Reply by phl0** [reply]
> Can the debugging information in the QSO list somehow be deactivated? It is not really an issue but if there is an easy way to do it .. ?

Set 
```
define('ENVIRONMENT', 'production');
```
in line 57 of index.php in the root web directory.

**Reply by AndreasK79** [reply]
> Regarding the user_id: We did have more than one user already in the old version, am I right? So how was it done there? Or was this data valid for all users before and is now on a per user base?
> 
> 
It was available for everyone. Now it filters on the logged in user.

**Reply by m0urs** [reply]
Thanks. So I would suggest that, if only one user exists, these stuff will be automatically added to that user. And if more than one user exists, it will be duplicated so that one entry exists for every user. In this case we would have the same behaviour as before? Otherwise this information will be lost (if you do not manually edit the database tables).

**Comment by magicbug** [comment]
Sorry I've been quiet all, June was a mess ended up in hospital then with COVID which completely wiped me out.

Hoping things area on the mend so I can get back to coding

**Comment by m0urs** [comment]
Thank you!

Andreas Kristiansen ***@***.***> schrieb am So. 26. Juni 2022
um 19:23:

> A lot has been changed in the database, but switching should not be a
> problem. I tried on my local machine to switch back to master, and it
> worked here.
>
> As always, make a backup, just in case. Better safe than sorry,
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1131#discussioncomment-3026933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATX5SUOE752XC3EA64XNOTVRCG2NANCNFSM5CCCR36Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
https://urspringer.de/info
Signal: <https://signal.org> +4915143106798
PGP-Verschlüsselung <https://www.gpg4win.de/>: https://urspringer.de/pgp

**Comment by electricEngineer** [comment]
I have a question regarding the promise of club-style logbooks mentioned in the first post. I just upgraded from a 1.X version to the latest one and now our club members can't access the club logbook anymore. Is there an option in any branch that you can share a logbook with all users on the clublog instance?

**Reply by HB9HIL** [reply]
Hello @electricEngineer 
unfortunately there isn't a feature like "Shared Logbooks" at the moment. 
We have this on our plan but no ETA. 
Sorry.
